docs/designs/CodeFlare-SDK-design-doc.md

```diff
@@ -0,0 +1,141 @@
+# CodeFlare SDK Design Document
+
+Team: Distributed Workloads - Orchestration
+
+## Introduction
+This document outlines the design of the Project CodeFlare SDK, a Python SDK that facilitates interactions between users and the distributed workloads component of Red Hat OpenShift AI(RHOAI)/ OpenDataHub(ODH). Users, in this instance, are both data scientists and MLOps Engineers. The SDK provides a high-level abstraction for managing machine learning(ML) workflows, jobs and distributed computing resources.
```

Comment:
```suggestion
This document outlines the design of the Project CodeFlare SDK, a Python SDK that facilitates interactions between users and the distributed workloads component of Red Hat OpenShift AI(RHOAI)/ OpenDataHub(ODH). Users, in this instance, are both data scientists and MLOps Engineers. The SDK provides a high-level abstraction for managing machine learning(ML) workflows, jobs and distributed computing resources.
```
Should this be more generic? Are we limiting our focus to RHOAI and ODH specifically?

Reply:
> Are we limiting our focus to RHOAI and ODH specifically?

Ideally the CodeFlare stack can be used on any k8s clusters (through ODH or standalone). But since it is being shipped with RHOAI and ODH, the SDK design mentions that the intention is to make interaction with those components easier. Is supporting CodeFlare components standalone something we are planning in future (also what would be the use case)? I believe as a data scientist I would prefer to adopt a whole ecosystem of components on my cluster (including NBs, IDE etc - though ODH), rather than just a piece of training stack. Wdyt?

Reply:
>I believe as a data scientist I would prefer to adopt a whole ecosystem of components on my cluster (including NBs, IDE etc - though ODH), rather than just a piece of training stack. Wdyt?

Candidly, as a data scientist I think I would want to use an SDK by a trusted source. the CodeFlare SDK is not well known among the community and it is a product we are advocating for customers to use. That's absolutely fine but I do think DS customers will view it differently than if it was from Kubeflow or KubeRay (ignoring the fact that we are actually making their lives easier). And I think that perception could influence adoption.

I don't think that should change what we do in the short term but it is a realistic gap in our current strategy; i.e., CodeFlare has a very limited community and limited brand awareness so it may face adoption challenges. Of course, we can see how this plays out with customers but I want to call it out.

Reply:
Completely fair, I feel we should track this somewhere. How about open an issue in here and also in JIRA - so that we can revisit when we get to having some of the codeFlare components upstream. 

Having it upstream would help us to easily cross-reference wherever required.